jax/experimental/sparse/transform.py

```diff
@@ -162,6 +163,97 @@ def argspec_to_aval(argspec):
   return tree_map(argspec_to_aval, argspecs, is_leaf=_is_argspec)
 
 
+#------------------------------------------------------------------------------
+# Implementation of sparsify() using tracers.
+
+class SparseTracer(core.Tracer):
+  def __init__(self, trace: core.Trace, *, argspec):
+    self._argspec = argspec
+    self._trace = trace
+
+  @property
+  def spenv(self):
+    return self._trace.main.spenv
+
+  @property
+  def aval(self):
+    return argspecs_to_avals(self.spenv, [self._argspec])[0]
+
+  def full_lower(self):
+    return self
+
+class SparseTrace(core.Trace):
+  def pure(self, val: Any):
+    argspec, = arrays_to_argspecs(self.main.spenv, [val])
+    return SparseTracer(self, argspec=argspec)
+
+  def lift(self, val: core.Tracer):
+    argspec, = arrays_to_argspecs(self.main.spenv, [val])
+    return SparseTracer(self, argspec=argspec)
+
+  def sublift(self, val: SparseTracer):
+    return SparseTracer(val._trace, argspec=val._argspec)
+
+  def process_primitive(self, primitive, tracers, params):
+    spenv = self.main.spenv
+    argspecs = [t._argspec for t in tracers]
+    if any(argspec.is_sparse() for argspec in argspecs):
+      if primitive not in sparse_rules:
+        raise NotImplementedError(f"sparse rule for {primitive}")
+      out_argspecs = sparse_rules[primitive](self.main.spenv, *(t._argspec for t in tracers), **params)
+    else:
+      out_bufs = primitive.bind(*(argspec.data(spenv) for argspec in argspecs), **params)
+      out_argspecs = arrays_to_argspecs(spenv, out_bufs if primitive.multiple_results else [out_bufs])
+    out_tracers = tuple(SparseTracer(self, argspec=argspec) for argspec in out_argspecs)
+    return out_tracers if primitive.multiple_results else out_tracers[0]
+
+  def process_call(self, call_primitive, f: lu.WrappedFun, tracers, params):
+    spenv = self.main.spenv
+    argspecs = tuple(t._argspec for t in tracers)
+    in_bufs = spenv._buffers
+    fun, out_argspecs = sparsify_subtrace(f, self.main, argspecs)
+    if any(params['donated_invars']):
+      raise ValueError("sparsify does not support donated_invars")
```

Comment:
```suggestion
      raise NotImplementedError("sparsify does not support donated_invars")
```

Just an idea. I think one way to support this would be to track how dense argument positions correspond to buffers underlying the sparse representations. Maybe we could implement that later, and hence a `NotImplementedError` could make sense.